### PR TITLE
Badge For Visual Studio Marketplace added.

### PR DESCRIPTION
**Pull request by ritwickdey**
Badge for Version
Version

Badge for Total No of Installs
Installs

Badge for Rating
rating

Issue Request https://github.com/badges/shields/issues/1067

### CODE REVIEW

**Comment by ritwickdey** [comment]
What is the error? 

`PR: badges/shields#1074 Error: 403 Forbidden`

**Comment by ritwickdey** [comment]
Build was falling due to https://github.com/badges/shields/issues/979 .

File: server.js [hunk 1]
```diff
@@ -4865,6 +4865,150 @@ cache(function(data, match, sendBadge, request) {
...
+        'content-type': 'application/json'
+      },
+      body:
+      {
+        filters: [{
+          criteria: [
+            { filterType: 7, value: repo },
+            { filterType: 12, value: '4096' }]
+        }],
+        flags: 914
```
Comment (paulmelnikow) [inline server.js]: What are these values? Could you add a comment?

Are the options identical between these three badges? If so could you make a `createVscodeOptions` to DRY this up?

File: service-tests/vscode.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+.expectJSONTypes(Joi.object().keys({
+  name: Joi.equal('Installs', 'installs'),
+  value: Joi.string().regex(/^[0-9]+[K|M]?$/i)
+}));
+
+t.create('rating')
+.get('/rating/ritwickdey.LiveServer.json')
+.expectJSONTypes(Joi.object().keys({
+  name: Joi.equal('rating', 'Rating'),
+  value: Joi.string()
```
Comment (paulmelnikow) [inline service-tests/vscode.js]: This is very broad, and matches e.g. the string `'undefined'`. What are the expected values here?

File: server.js (hunk 1 above)
Comment (ritwickdey) [reply server.js]: Yep identical! this is for API request. Okay I will separate it. 😃

File: service-tests/vscode.js (hunk 2 above)
Comment (ritwickdey) [reply service-tests/vscode.js]: Ok! I'll add validation

**Comment by ritwickdey** [comment]
Hi, @paulmelnikow  I've separated common code and fixed a test case.

File: service-tests/vscode.js [hunk 3]
```diff
@@ -6,21 +6,21 @@ const ServiceTester = require('./runner/service-tester');
...
+t.create('rating should be formatted. eg. 4.25/5 (30)')
 .get('/rating/ritwickdey.LiveServer.json')
 .expectJSONTypes(Joi.object().keys({
   name: Joi.equal('rating', 'Rating'),
-  value: Joi.string()
+  value: Joi.string().regex(/[0-5].[0-9]{2}\/5?\s*\([0-9]*\)$/)
 }));
 
-t.create('version')
+t.create('version should be formatted. eg. v7.2.5')
```
Comment (paulmelnikow) [inline service-tests/vscode.js]: Love these.

File: service-tests/vscode.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+'use strict';
+
+const Joi = require('joi');
+const ServiceTester = require('./runner/service-tester');
+
+const t = new ServiceTester({ id: 'vscode', title: 'VS Code Marketplace' });
+module.exports = t;
+
+t.create('installs should be formatted. eg. 72M')
+.get('/installs/ritwickdey.LiveServer.json')
```
Comment (paulmelnikow) [inline service-tests/vscode.js]: Could you change `installs` to `d`, `version` to `v`, and `rating` to `r` for consistency? Sorry I missed this before!

File: server.js [hunk 5]
```diff
@@ -4865,32 +4865,48 @@ cache(function(data, match, sendBadge, request) {
...
+function getVscodeStatistic(data, statisticName) {
+  let statistics = data.results[0].extensions[0].statistics;
+  try {
+    let statistic = statistics.find(x => x.statisticName.toLowerCase() === statisticName.toLowerCase());
+    return statistic.value;
+  } catch (err) {
+    return 0; //In case required statistic is not found means ZERO.
+  }
+}
+
```
Comment (paulmelnikow) [inline server.js]: Thanks for refactoring this!

File: service-tests/vscode.js (hunk 4 above)
Comment (ritwickdey) [reply service-tests/vscode.js]: Okay. no problem.. I'll :smile:

**Comment by paulmelnikow** [comment]
Thinking this through a little more, would `/vscode-marketplace` be a clearer URL path?

**Comment by ritwickdey** [comment]
Yep, conventionally `/vscode-marketplace`  is more cleaner route.. But I thought, it's too big route. Suggest me what you would like to prefer. :smile:

[I am totally newbie, I'll do what you suggest me] :smile:

**Comment by paulmelnikow** [comment]
I don't think it's too long. 😄

**Comment by ritwickdey** [comment]
:smile: Okay. No Problem (Thanks for the suggestion).. I am now out of home. As soon as I reach to my home, I'll change all the url path. :smile:

**Comment by ritwickdey** [comment]
Hi @paulmelnikow, I have a question. If I want to add download badge in 3 different format (eg 7K, 7.45K or 7458). What should be routing? Should I add 3 different format?

**Comment by paulmelnikow** [comment]
I'd prefer to use our standard number formatting using `metric()`, which provides a more consistent experience across different badges. I've made that argument in #999, e.g.

**Comment by ritwickdey** [comment]
yep! I used the `metric()` function. Anyway. please review my code. I've changed the route accordingly and have done small refactoring. 😃

File: server.js [hunk 6]
```diff
@@ -4865,6 +4865,92 @@ cache(function(data, match, sendBadge, request) {
...
+          }
+          case 'r': {
+            badgeData.text[0] = getLabel('rating', data);
+            let rate = getVscodeStatistic(buffer, 'averagerating').toFixed(2);
+            let totalrate = getVscodeStatistic(buffer, 'ratingcount');
+            badgeData.text[1] = rate + '/5 (' + totalrate + ')';
+            break;
+          }
+          case 'v': {
+            badgeData.text[0] = getLabel('Visual Studio Marketplace', data);
```
Comment (paulmelnikow) [inline server.js]: Could you make this lower case, like we do for the other app stores? Sorry I didn't notice it before.

Comment (ritwickdey) [reply server.js]: You mean that the label 'Visual Studio Marketplace' to 'visual studio marketplace'. Right?

Comment (paulmelnikow) [reply server.js]: Correct!

**Comment by ritwickdey** [comment]
Changed to lower case 😊

**Comment by paulmelnikow** [comment]
Thanks a bunch!

**Comment by ritwickdey** [comment]
Thanks a lot to you. 😃

**Pull request merged**